### PR TITLE
Replacing the old save control in the calculator with a new save control

### PR DESCRIPTION
**Pull request by Wycklife**
@africanmathsinitiative/developers This is work in progress
I have added the new control to the calculator. It adds a new column to the data frame. The part remaining is positioning of the new column.

### CODE REVIEW

**Comment by Wycklife** [comment]
@africanmathsinitiative/developers This is now ready for review. I tried looking at the dialogs which uses the save control, I keep on getting the bug below whenever I reopen the dialog after positioning the new column. @lloyddewit What could be the cause of this bug?

image

**Comment by lloyddewit** [comment]
@Wycklife I assume the error is because the data name is not provided (it's an empty string). However, I couldn't recreate the error. I tested the PR and it seems to do the column positioning correctly :). See screen dump below.
There are still some issues related to updating the prefix.
If you have specific steps that trigger the error then let me know and I'll try and recreate.
image

**Comment by Wycklife** [comment]
@lloyddewit Can the error be caused by machine? I also tested the same using Ivan's laptop and it never produced the error. Let me look into it a gain. I'm also fixing the prefix

**Comment by lloyddewit** [comment]
@Wycklife Interesting. The error is most likely caused by slightly different steps being taken on the 2 laptops. Or it could be a difference in the environment such as OS, dotnet version, Visual Studio etc.. Or it could be problems initialising the memory location used to store/update the data name (especially if >1 variables reference the same location), but this is a more common problem with older languages and VB dotnet has memory management features to prevent this type of issue.

**Comment by Patowhiz** [comment]
> @africanmathsinitiative/developers This is now ready for review. I tried looking at the dialogs which uses the save control, I keep on getting the bug below whenever I reopen the dialog after positioning the new column. @lloyddewit What could be the cause of this bug?
> 
> image

@Wycklife from you comment above. From the error message box it is quite clear that the data name is not being found. It's indicated as empty "". Thus the data object is not also found. I think this could be caused by how your code is set up on reopening the dialog. 

I'm happy to review this and get more insights on the code.

**Comment by rdstern** [comment]
@Patowhiz your review would be good.  @Wycklife is currently main preparing for his MSc exams, so I am not sure when he will look at the review, but it would be good to do antyway.

**Comment by Wycklife** [comment]
@Patowhiz This error occurs even on the count in factor dialog not only on the calculator, I thought perhaps it's caused by my machine, of which I agree with @lloyddewit  comment above. I tried testing these dialogs using Ivan's laptop but I couldn't find the error, using the same procedure I used on with my laptop. @lloyddewit  tested too and never found the error.

**Comment by Wycklife** [comment]
@Patowhiz Did you try to test this?

File: instat/dlgCalculator.vb [hunk 1]
```diff
@@ -41,10 +41,10 @@ Public Class dlgCalculator
...
-            If ucrCalc.chkSaveResultInto.Checked AndAlso ucrCalc.ucrSaveResultInto.IsEmpty Then
-                ucrBase.OKEnabled(False)
-            Else
+            If ucrCalc.ucrSaveResultInto.IsComplete Then
                 ucrBase.OKEnabled(True)
+            Else
+                ucrBase.OKEnabled(False)
             End If
         Else
             ucrBase.OKEnabled(False)
```
Comment (Patowhiz) [inline instat/dlgCalculator.vb:50]: @Wycklife  The whole code statement inside TestOkEnabled() subroutine could be replaced with 

` ucrBase.OKEnabled(Not ucrCalc.ucrReceiverForCalculation.IsEmpty AndAlso ucrCalc.ucrSaveResultInto.IsComplete)`

File: instat/ucrCalculator.vb [hunk 2]
```diff
@@ -983,7 +983,7 @@ Public Class ucrCalculator
         End If
     End Sub
 
-    Private Sub chkSaveResultInto_CheckedChanged(sender As Object, e As EventArgs) Handles chkSaveResultInto.CheckedChanged
+    Private Sub chkSaveResultInto_CheckedChanged(sender As Object, e As EventArgs)
```
Comment (Patowhiz) [inline instat/ucrCalculator.vb:986]: @Wycklife  Do we still need this subroutine? I think this was meant to be an event handler for  control `chkSaveResultInto` which has been removed

File: instat/ucrCalculator.vb [hunk 3]
```diff
@@ -1230,7 +1230,7 @@ Public Class ucrCalculator
         RaiseEvent SelectionChanged()
     End Sub
 
-    Private Sub ucrSaveResultInto_NameChanged() Handles ucrSaveResultInto.NameChanged
+    Private Sub ucrSaveResultInto_NameChanged()
```
Comment (Patowhiz) [inline instat/ucrCalculator.vb:1233]: @Wycklife I couldn't find where `ucrSaveResultInto_NameChanged()` is used. Even event `SaveNameChanged()` is no longer being raised.

Comment (Patowhiz) [reply instat/ucrCalculator.vb:1233]: It shows up as unused when I looked it up in visual studio

**Comment by Patowhiz** [comment]
@wycliffe I also noted that there is a big space between the base buttons and `ucrSave` control.

calculator

**Comment by Wycklife** [comment]
> @Wycklife in addition to the queries I have raised I could not build this PR without commenting out the usage of `ucrCalculator` in `sdgCalculationsSummary`. You may want to check the changes required before this is merged.
> 
> Below is the `sdgCalculationsSummary` that I had to comment out.
> sdgCalc

@Patowhiz This should be fixed on a separate branch, it's a different sub dialog using the same old save control. What do you think? Should I remove it on this same branch

File: instat/ucrCalculator.vb (hunk 3 above)
Comment (Wycklife) [reply instat/ucrCalculator.vb:1233]: > @Wycklife I couldn't find where `ucrSaveResultInto_NameChanged()` is used. Even event `SaveNameChanged()` is no longer being raised.

@Patowhiz Could you check a gain please?

**Comment by Wycklife** [comment]
@Patowhiz Could you review this a gain. I have fixed the above comments.

**Comment by Wycklife** [comment]
@lloyddewit I have made improvements to this pull request and it seems okay from my end.

**Comment by Wycklife** [comment]
@lloyddewit The question I have is whether to remove that line of code from sdgCalucaltionSummary which produces an error when building. If yes then should I do it on this branch?

**Comment by lloyddewit** [comment]
@Wycklife yes, you should delete the line and include the change in this PR.
If you submit a PR then that PR should be buildable and testable. In this PR, you deleted `chkSaveResultInto`. Therefore you also need to remove any `chkSaveResultInto` references in other files. Otherwise the software will not build. All changed files should be included in the PR. I don't understand how you were able to build and test this PR without removing this line. Were you using an older version of `sdgCalculationSummary`?

**Comment by Wycklife** [comment]
> @Wycklife yes, you should delete the line and include the change in this PR.
> If you submit a PR then that PR should be buildable and testable. In this PR, you deleted `chkSaveResultInto`. Therefore you also need to remove any `chkSaveResultInto` references in other files. Otherwise the software will not build. All changed files should be included in the PR. I don't understand how you were able to build and test this PR without removing this line. Were you using an older version of `sdgCalculationSummary`?

@lloyddewit Thanks for your comment. I think whenever I build, then this error keeps on popping up, so I had to comment the error out. Let me remove that part.

**Comment by Wycklife** [comment]
@lloyddewit I have removed the error.

**Comment by lloyddewit** [comment]
@Patowhiz Please could you confirm that your comments are resolved? thanks

**Comment by Patowhiz** [comment]
@Wycklife the code looks good to me.
There are just 2 lines of code that define variables which are no longer being used. Code line 21 and  22
```
 Dim strCalcHistory As List(Of String)
 Dim dataset As DataFrame
```
I think the 2 lines can be deleted, they add no functionality to the dialog.

On the design, I can still see a huge gap between the save control and the base buttons, I wonder if this is happening on my screen only? I checked and didn't see any hidden control occupying that space.

space dlgenter

**Comment by Wycklife** [comment]
@Patowhiz Some of the keyboards are large in size i.e Wakefield keyboard, that's why that space  was left. It has been there even before.

**Comment by rdstern** [comment]
@Wycklife you have a point, but here is the dialogue with Wakefield.

**Comment by rdstern** [comment]
@Wycklife you have a point.  But here is the calculator with the Wakefield keyboard.  So I still think that, as @Patowhiz suggests, the gap doesn't need to be so large?

image

**Comment by Wycklife** [comment]
@rdstern I have reduced the space, I just want to commit the changes.

**Comment by rdstern** [comment]
@Wycklife that's great.  It was a good suggestion from @Patowhiz .  Once you reduced the space I assume (and hope) that you checked with the big keyboards - especially Wakefield.  If you remember we found earlier that with Wakefield the dialogue looked odd when you returned to the dialogue a second time.

**Comment by Wycklife** [comment]
@Patowhiz , @rdstern I suggest we leave that space the way it is. I have tried reducing the size but it's affecting the wakefield keyboard when you reopen the dialog. I.e Some of the keys are hidden.

**Comment by Patowhiz** [comment]
@Wycklife can you please share the screenshot.

**Comment by Patowhiz** [comment]
@Wycklife  If it's retaining the width then I think it should retain the height too on reopening. Generally VB.NET handles the forms and controls size measurements by handling the 2 in it's events. we can have a quick skype about it

**Comment by rdstern** [comment]
Aha, @Wycklife am I right that when you reduce the space it is ok when you first go into the calculator and then use Wakefield?  But then, when you return to the calculator again it isn't ok?

Or is it a problem anyway?

**Comment by Wycklife** [comment]
> Aha, @Wycklife am I right that when you reduce the space it is ok when you first go into the calculator and then use Wakefield? But then, when you return to the calculator again it isn't ok?
> 
> Or is it a problem anyway?

Yes, you're right. I am now setting the height of the dialog in relation with the keyboard selected. Just as the width of the dialog enlarges or reduces depending on the size of the keyboard selected. I think this will work out.

**Comment by Wycklife** [comment]
@rdstern This is now working well on my side. The location of the ucrBase and the height of the dialog changes depending on the keyboard option selected. @Patowhiz Could you please review if this is okay? @rdstern You can review as well. Thank you.

**Review by rdstern — CHANGES_REQUESTED** [review]
@Wycklife the save part now seems to work.  But on the layout:
a) The Comment field, in particular, has now moved very close to the left-hand side.  I was happy where it was before.
b) With the Wakefield keyboard I don't see why the left hand side needs to change.  There seems to be a lot of space with the left side just remaining where it is.  Are there any other keyboards that change the layout?  Do they need to?

**Comment by Patowhiz** [comment]
@Wycklife I have made changes to the code that calculates the required height to your branch in PR [204](https://github.com/Wycklife/R-Instat/pull/204). You can merge to your branch. I cloned my branch from this PR branch.

Below is a screenshot of the new design with and without the wakefield calculator visible.

Without wakefield calculator
basic calculator

With wakefield calculator
wakefield calc

**Comment by lloyddewit** [comment]
@Wycklife is this PR now ready for review/test or do you still expect more commits? thanks for clarifying

**Comment by Wycklife** [comment]
@rdstern The following are the changes that @Patowhiz  made.
1. Maitined the height of the dialog and the location of the base control.
2. Changes the width of the dialog to accommodate the calculator size.
@lloyddewit This now looks okay. @africanmathsinitiative/developers Someone to review the changes.

**Comment by Wycklife** [comment]
@rdstern The following are the changes that @Patowhiz  made.
1. Maitined the height of the dialog and the location of the base control.
2. Changes the width of the dialog to accommodate the calculator size.
@lloyddewit This now looks okay. @africanmathsinitiative/developers Someone to review the changes.

**Comment by Wycklife** [comment]
@africanmathsinitiative/developers This branch has got conflict, am fixing them now

**Comment by Wycklife** [comment]
@rdstern I realised some of the keyboard options were hidden and I have made some improvements to them. I have also resolve the conflicts for this pr. Could you please review this once more? I might not be able to continue working on the date keyboard till this pr is merged to avoid conflicts. Thank you.

**Comment by Wycklife** [comment]
@lloyddewit This PR is now ready for review. Thank you.

**Comment by rdstern** [comment]
Thanks @Wycklife and could you specify which keyboards etc you changed.  There is a lot in this dialogue and I would like to be sure I am checking exactly the bits you changed.

**Comment by Wycklife** [comment]
> Thanks @Wycklife and could you specify which keyboards etc you changed. There is a lot in this dialogue and I would like to be sure I am checking exactly the bits you changed.

These are the keyboards I made improvements:
Summary Keyboard
Strings (Character Columns)
Wakefield
Transform

File: instat/dlgCalculator.vb [hunk 4]
```diff
@@ -1,194 +1,172 @@
...
+    Private Sub ucrBase_ClickReset(sender As Object, e As EventArgs) Handles ucrBase.ClickReset
+        SetDefaults()
+        TestOKEnabled()
+    End Sub
+
+    Private Sub ucrInputCalOptions_NameChanged() Handles ucrCalc.NameChanged
+        CalculationsOptions()
+    End Sub
+
+    Private Sub CalculationsOptions()
```
Comment (Patowhiz) [inline instat/dlgCalculator.vb:133]: Probably a brief comment that explains what `CalculationsOptions()` does.

File: instat/ucrCalculator.vb (hunk 3 above)
Comment (Patowhiz) [reply instat/ucrCalculator.vb:1235]: Still shows up as unused

File: instat/ucrCalculator.vb (hunk 2 above)
Comment (Patowhiz) [reply instat/ucrCalculator.vb:988]: Same here. Unused subroutine

**Comment by rdstern** [comment]
As you are on this dialogue, please can I add a (hopefully trivial) addition.  It is one that could affect many dialogues and good that we take the opportunity to correct, when we are changing a dialogue.

There is a tidyverse style guide.  It says:
"Variable and function names should use **_only lowercase letters, numbers, and __**. Use underscores (_) (so called snake case) to separate words within a name.


# Good
day_one
day_1

# Bad
DayOne
dayone"

In the calculate dialogue we suggest a new variable is Calc1, i.e. with upper case C.  Please change to calc1.  Thanks.

**Comment by Patowhiz** [comment]
Functionality and code looks good from my end.

**Review by rdstern — DISMISSED** [review]
Happy to approve.  Over to @lloyddewit .  If ok, good to include in the next release

File: instat/dlgCalculator.vb [hunk 5]
```diff
@@ -1,194 +1,172 @@
...
+            ucrBase.clsRsyntax.SetAssignTo(ucrCalc.ucrSaveResultInto.GetText(), strTempColumn:=ucrCalc.ucrSaveResultInto.GetText(), strTempDataframe:=ucrCalc.ucrSelectorForCalculations.ucrAvailableDataFrames.cboAvailableDataFrames.Text)
+            ucrBase.clsRsyntax.bExcludeAssignedFunctionOutput = True
+            ucrBase.clsRsyntax.iCallType = 0
+        Else
+            ucrBase.clsRsyntax.RemoveAssignTo()
+            ucrBase.clsRsyntax.iCallType = 2
+            ucrBase.clsRsyntax.bExcludeAssignedFunctionOutput = False
+        End If
+    End Sub
+    Private Sub ucrBase_ClickOk(sender As Object, e As EventArgs) Handles ucrBase.ClickOk
```
Comment (lloyddewit) [inline instat/dlgCalculator.vb:114]:
```suggestion

    Private Sub ucrBase_ClickOk(sender As Object, e As EventArgs) Handles ucrBase.ClickOk
```

File: instat/dlgCalculator.vb [hunk 6]
```diff
@@ -1,194 +1,172 @@
...
+    Private Sub ucrBase_ClickReset(sender As Object, e As EventArgs) Handles ucrBase.ClickReset
+        SetDefaults()
+        TestOKEnabled()
+    End Sub
+
+    Private Sub ucrInputCalOptions_NameChanged() Handles ucrCalc.NameChanged
+        CalculationsOptions()
+    End Sub
+
+    Private Sub CalculationsOptions()
```
Comment (lloyddewit) [inline instat/dlgCalculator.vb:133]:
```suggestion
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]

**Review by lloyddewit — APPROVED** [review]
@Wycklife Thank you, this looks good. I just made a couple of small cosmetic changes.

**Comment by lloyddewit** [comment]
@rdstern Please could you re-approve (functionality has not changed)? thanks

**Review by rdstern — APPROVED** [review]

**Pull request merged**